src/main/java/org/apache/sling/models/impl/injectors/OSGiServiceInjector.java

```diff
@@ -204,7 +204,11 @@ public Callback(ServiceReference<?>[] refs, BundleContext context) {
         public void onDisposed() {
             if (refs != null) {
                 for (ServiceReference<?> ref : refs) {
-                    context.ungetService(ref);
+                    try {
+                        context.ungetService(ref);
+                    } catch (IllegalStateException | IllegalArgumentException | NullPointerException exception) {
```

Comment:
Please let me know if I should replace multiple exception with a generic RuntimeException or this should be handled in a different way

Reply:
NPE should never been caught. That hints at a mistake in the code. Same for IAE according to https://docs.osgi.org/javadoc/osgi.core/7.0.0/org/osgi/framework/BundleContext.html#ungetService-org.osgi.framework.ServiceReference-. Only catching ISE might make sense here, although it is not really clear to me how this can happen....

Reply:
Hi Kwin, Thanks for the comment. I agree that NPE, IAE or any other runtime exception should not be caught unless we can recover safely from it. But point here it to let the dispose the all references present in the queue [0] at once (if present). I will remove the NPE and IAE from catch block. I have tried to explain the scenario in JIRA [1].

Whenever we use a model (let's a M in Bundle B) with OSGi service injection and use the model in a page.. OSGiServices are injected into the model and for disposing these injected references ModelAdapterFactory keeps PhantomReferences with reference queue (see ModelAdapterFactory for more details [2]). This queue is checked every 30 seconds (default), to disposes these references. And if Bundle B is refreshed (restarted) before the references were disposed, while disposing these references (context.ungetService(ref)) we get ISE: Invalid Bundle Context. which stops the current job trigger and the next reference is picked in the next job trigger.

Please let me If I am missing something.

[0]: https://github.com/apache/sling-org-apache-sling-models-impl/blob/master/src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java#L210
[1]: https://issues.apache.org/jira/browse/SLING-11132
[2]: https://github.com/apache/sling-org-apache-sling-models-impl/blob/master/src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java

Reply:
Just unget the service reference in `deactivate` and at the same time remove it from the reference queue.

Reply:
I think it is a valid case to catch all these RuntimeExceptions to guarantee that the iteration over the ServiceReference completes. However, as you don't handle the exceptions you may want to collect them instead and add them as suppressed exceptions to a new `RuntimeException` that you throw if any exception happened during the iteration.

Reply:
Thank you everyone for guidance, update the code. Please let me know if anything else if also required